### PR TITLE
[fix][torch.compile] Fix cold-start compilation time increase by adding kv cache update to splitting ops

### PR DESCRIPTION
**Pull request by ProExpertProg**
## Purpose
As described in #33267, a cold start time regression was introduced by #25954. This fixes that (temporarily) by adding `unified_kv_cache_update` to `splitting_ops`. The PR also includes a fix to make sure `unified_kv_cache_update` and `unified_attention` end up in the same splitting graph to reduce piecewise backend overhead.

## Startup Results
Numbers for compile/total startup time (clearly our compilation time reporting is imperfect), on B200 using `TRITON_ATTN` (to trigger separate `unified_kv_cache_update`):

perf tl;dr: if there's a perf regression, I wasn't able to find it.

### This PR
- gpt-oss-120b tp=4:
  - cold start: 13.44s / 72.35s
  - warm start: 7.38s / 66.90s
  - median latency: 0.5260136430151761s, 0.5177223119972041s, 0.5080516894959146s
- llama3-70b:
  - cold start: 22.03s / 76.90s 
  - warm start: 15.40s / 62.66s
  - median latency: 3.0474346815026365s
  - median latency (commit 2): 3.0955106309993425s
- llama3-8b (H100):
  - median latency: 0.8313153050839901s
  - median latency (commit 2): 0.8302149870432913s

### Main
- gpt-oss-120b tp=4:
  - cold start: 43.12s / 102.79s
  - warm start: 7.25s / 66.34s
  - median latency: 0.5062499015039066s, 0.5127717430150369s, 0.5170540044928202s
- llama3-70b:
  - cold start: 87.56s / 141.71s
  - warm start: 14.79s / 60.08s
  - median latency: 3.0957505205005873s
- llama3-8b (H100):
  - median latency: 0.8309728200547397s

### This PR

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.46|±  |0.0501|
|     |       |strict-match    |     5|exact_match|↑  | 0.35|±  |0.0479|

### This PR (inductor partition, op not excluded)
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.50|±  |0.0503|
|     |       |strict-match    |     5|exact_match|↑  | 0.34|±  |0.0476|

### This PR (manually set splitting ops to not include kv cache)
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.46|±  |0.0501|
|     |       |strict-match    |     5|exact_match|↑  | 0.35|±  |0.0479|

### Main
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.42|±  |0.0496|
|     |       |strict-match    |     5|exact_match|↑  | 0.24|±  |0.0429|

## Test Plan
CI, quick llama3 eval

## Test Result
Looks good

### CODE REVIEW

**Comment by zou3519** [comment]
@ProExpertProg I'm wondering about runtime overhead now (esp for piecewise cudagraphs). My understanding is that after this PR the kv_cache_update will be in a separate submodule than the attention operation (I haven't run it though, if it's in the same submodule then we're good). Having the separate subgraph leads to a bit more runtime overhead (I'm not sure how much).

@BoyuanFeng did some benchmarking in the past where he noticed part of the speed up of Inductor graph partition over the FX graph partition was in straight up overhead reduction from not needing to run submodules in the FX interpreter.

**Comment by ProExpertProg** [comment]
It does seem like there's a minor (4%) regression in latency for gpt-oss-120B: 0.5260136430151761/0.5062499015039066 = 3.9%.

Let me try the "mild revert" approach where kv cache goes back inside the attention op

**Comment by ProExpertProg** [comment]
I have manually verified that the new change produces a single splitting graph:
```
          
         # File: /home/ProExpertProg/git/vllm/vllm/model_executor/layers/attention/attention.py:424 in forward, code: value = value.view(-1, self.num_kv_heads, self.head_size_v)
        view_5: "bf16[s72, 2, 64]" = contiguous.view(-1, 2, 64);  contiguous = None
        return (view_4, view_5, view_2, view_3, all_reduce)
        
class submod_1(torch.nn.Module):
    def forward(self, key_2: "bf16[s72, 2, 64]", s72: "Sym(s72)", value: "bf16[s72, 2, 64]", query_2: "bf16[s72, 16, 64]", output_4: "bf16[s72, 16, 64]"):
         # File: /home/ProExpertProg/git/vllm/vllm/model_executor/layers/attention/attention.py:445 in forward, code: kv_cache_dummy_dep = torch.ops.vllm.unified_kv_cache_update(
        unified_kv_cache_update: "bf16[0]" = torch.ops.vllm.unified_kv_cache_update(key_2, value, 'model.layers.0.attn.attn')
        
         # File: /home/ProExpertProg/git/vllm/vllm/model_executor/layers/attention/attention.py:448 in forward, code: torch.ops.vllm.unified_attention_with_output(
        unified_attention_with_output = torch.ops.vllm.unified_attention_with_output(query_2, key_2, value, output_4, 'model.layers.0.attn.attn', kv_cache_dummy_dep = unified_kv_cache_update);  query_2 = key_2 = value = output_4 = unified_kv_cache_update = unified_attention_with_output = None
        return ()
        
class submod_2(torch.nn.Module):
    def forward(self, output_4: "bf16[s72, 16, 64]", s72: "Sym(s72)", l_self_modules_layers_modules_0_modules_attn_modules_o_proj_parameters_weight_: "bf16[2880, 1024]", l_self_modules_layers_modules_0_modules_attn_modules_o_proj_parameters_bias_: "bf16[2880]", l_self_modules_layers_modules_0_modules_post_attention_layernorm_parameters_weight_: "bf16[2880]", output: "bf16[s72, 2880]", l_self_modules_layers_modules_0_modules_mlp_modules_router_parameters_weight_: "bf16[128, 2880]", l_self_modules_layers_modules_0_modules_mlp_modules_router_parameters_bias_: "bf16[128]", l_self_modules_layers_modules_1_modules_input_layernorm_parameters_weight_: "bf16[2880]", l_self_modules_layers_modules_1_modules_attn_modules_qkv_proj_parameters_weight_: "bf16[1280, 2880]", l_self_modules_layers_modules_1_modules_attn_modules_qkv_proj_parameters_bias_: "bf16[1280]", l_positions_: "i64[s72]", l_self_modules_layers_modules_0_modules_attn_modules_rotary_emb_buffers_cos_sin_cache_: "f32[131072, 64]"):
         # File: /home/ProExpertProg/git/vllm/vllm/model_executor/layers/attention/attention.py:456 in forward, code: return output.view(-1, hidden_size)
        view: "bf16[s72, 1024]" = output_4.view(-1, 1024);  output_4 = None
                  
```

**Review by zou3519 — APPROVED** [review]
lgtm, write a unittest for splitting if you can.

**Comment by zou3519** [comment]
@ProExpertProg here's an e2e test if you need to rebase this: https://gist.github.com/zou3519/5e4478593a8f14ff1768b14e15db50eb

otherwise I can check it in after this goes in

**Review by mgoin — APPROVED** [review]
